File: go-RayTracer/camera.go
package main

import (
	"fmt"
	"io"
	"math"
	"os"
	"sync"
)

// Camera represents the view into the scene
type Camera struct {
	// Public configuration parameters
	AspectRatio     float64
	ImageWidth      int
	SamplesPerPixel int
	MaxDepth        int
	Vfov            float64
	LookFrom        Point3
	LookAt          Point3
	Vup             Vec3
	DefocusAngle    float64
	FocusDist       float64

	// Private fields
	imageHeight       int
	pixelSamplesScale float64
	center            Point3
	pixel00Loc        Point3
	pixelDeltaU       Vec3
	pixelDeltaV       Vec3
	u, v, w           Vec3
	defocusDiskU      Vec3
	defocusDiskV      Vec3
}

// NewCamera creates a camera with default settings
func NewCamera() *Camera {
	return &Camera{
		AspectRatio:     1.0,
		ImageWidth:      100,
		SamplesPerPixel: 10,
		MaxDepth:        10,
		Vfov:            90,
		LookFrom:        Point3{[3]float64{0, 0, 0}},
		LookAt:          Point3{[3]float64{0, 0, -1}},
		Vup:             Vec3{[3]float64{0, 1, 0}},
		DefocusAngle:    0,
		FocusDist:       10,
	}
}

// Initialize prepares the camera for rendering
func (c *Camera) Initialize() {
	c.imageHeight = int(float64(c.ImageWidth) / c.AspectRatio)
	if c.imageHeight < 1 {
		c.imageHeight = 1
	}

	c.pixelSamplesScale = 1.0 / float64(c.SamplesPerPixel)
	c.center = c.LookFrom

	// Determine viewport dimensions
	theta := DegreesToRadians(c.Vfov)
	h := math.Tan(theta / 2)
	viewportHeight := 2 * h * c.FocusDist
	viewportWidth := viewportHeight * (float64(c.ImageWidth) / float64(c.imageHeight))

	// Calculate the u,v,w unit basis vectors for the camera coordinate frame
	c.w = UnitVector(c.LookFrom.Sub(c.LookAt))
	c.u = UnitVector(Cross(c.Vup, c.w))
	c.v = Cross(c.w, c.u)

	// Calculate vectors across the horizontal and down the vertical viewport edges
	viewportU := c.u.MulScalar(viewportWidth)
	viewportV := c.v.MulScalar(-viewportHeight)

	// Calculate the horizontal and vertical delta vectors from pixel to pixel
	c.pixelDeltaU = viewportU.DivScalar(float64(c.ImageWidth))
	c.pixelDeltaV = viewportV.DivScalar(float64(c.imageHeight))

	// Calculate the location of the upper left pixel
	viewportUpperLeft := c.center.
		Sub(c.w.MulScalar(c.FocusDist)).
		Sub(viewportU.DivScalar(2)).
		Sub(viewportV.DivScalar(2))
	c.pixel00Loc = viewportUpperLeft.Add(
		c.pixelDeltaU.Add(c.pixelDeltaV).MulScalar(0.5))

	// Calculate the camera defocus disk basis vectors
	defocusRadius := c.FocusDist * math.Tan(DegreesToRadians(c.DefocusAngle/2))
	c.defocusDiskU = c.u.MulScalar(defocusRadius)
	c.defocusDiskV = c.v.MulScalar(defocusRadius)
}

// GetRay returns a ray for a given pixel coordinate
func (c *Camera) GetRay(i, j int) Ray {
	// Construct a camera ray originating from the defocus disk and directed at a randomly
	// sampled point around the pixel location i, j
	offset := c.sampleSquare()
	pixelSample := c.pixel00Loc.
		Add(c.pixelDeltaU.MulScalar(float64(i) + offset.X())).
		Add(c.pixelDeltaV.MulScalar(float64(j) + offset.Y()))

	rayOrigin := c.center
	if c.DefocusAngle > 0 {
		rayOrigin = c.defocusDiskSample()
	}
	rayDirection := pixelSample.Sub(rayOrigin)

	return Ray{Orig: rayOrigin, Dir: rayDirection}
}

// Render renders the scene with the given world
func (c *Camera) Render(world Hittable, out io.Writer, numThreads int) {
	c.Initialize()

	// Create image data
	img := NewImage(c.ImageWidth, c.imageHeight)

	if numThreads == 1 {
		for j := range c.imageHeight {
			fmt.Fprintf(os.Stdout, "\rScanlines remaining: %d ", c.imageHeight-j)
			for i := range c.ImageWidth {
				pixelColor := Color{[3]float64{0, 0, 0}}
				for range c.SamplesPerPixel {
					r := c.GetRay(i, j)
					pixelColor = pixelColor.Add(c.rayColor(r, c.MaxDepth, world))
				}
				img.SetPixel(i, j, pixelColor.MulScalar(c.pixelSamplesScale))
			}
		}
	} else {
		// Add a maximum of goroutines to the pool
		// This is a simple way to limit the number of concurrent goroutines
		// without using a channel or sync package

		// Create a wait group to synchronize goroutines
		var wg sync.WaitGroup
		waitChan := make(chan struct{}, numThreads)
		lines_remaining := c.imageHeight

		for pixel_idx := range c.imageHeight * c.ImageWidth {
			waitChan <- struct{}{}
			wg.Add(1)
			go func(pixel_idx int) {
				defer wg.Done()
				i := pixel_idx % c.ImageWidth
				j := pixel_idx / c.ImageWidth
				pixelColor := Color{[3]float64{0, 0, 0}}
				for range c.SamplesPerPixel {
					r := c.GetRay(i, j)
					pixelColor = pixelColor.Add(c.rayColor(r, c.MaxDepth, world))
				}
				img.SetPixel(i, j, pixelColor.MulScalar(c.pixelSamplesScale))
				if i == 0 {
					lines_remaining--
					fmt.Fprintf(os.Stdout, "\rScanlines remaining: %d ", lines_remaining)
				}
				<-waitChan
			}(pixel_idx)
		}
		wg.Wait()
	}
	fmt.Fprintf(os.Stdout, "\rScanlines remaining: 0 ")

	// Write the image to provided output
	img.WriteToStream(out)

	fmt.Fprintf(os.Stdout, "\rDone.                 \n")
}

// sampleSquare returns a random point in the [-0.5,0.5]x[-0.5,0.5] unit square
func (c *Camera) sampleSquare() Vec3 {
	return Vec3{[3]float64{
		RandomDouble() - 0.5,
		RandomDouble() - 0.5,
		0,
	}}
}

// defocusDiskSample returns a random point in the camera defocus disk
func (c *Camera) defocusDiskSample() Point3 {
	p := RandomInUnitDisk()
	return c.center.
		Add(c.defocusDiskU.MulScalar(p.E[0])).
		Add(c.defocusDiskV.MulScalar(p.E[1]))
}

// rayColor determines the color seen along a ray
func (c *Camera) rayColor(r Ray, depth int, world Hittable) Color {
	// If we've exceeded the ray bounce limit, no more light is gathered
	if depth <= 0 {
		return Color{[3]float64{0, 0, 0}}
	}

	rec := HitRecord{}
	if world.Hit(r, Interval{0.001, Infinity}, &rec) {
		reflect, attenuation, scattered := rec.Mat.Scatter(r, rec)
		if reflect {
			return attenuation.Mul(c.rayColor(scattered, depth-1, world))
		}
		return Color{[3]float64{0, 0, 0}}
	}

	unitDirection := UnitVector(r.Direction())
	a := 0.5 * (unitDirection.Y() + 1.0)
	white := Color{[3]float64{1.0, 1.0, 1.0}}
	blue := Color{[3]float64{0.5, 0.7, 1.0}}
	return white.MulScalar(1.0 - a).Add(blue.MulScalar(a))
}

File: ts-RayTracer/src/camera.ts
import { degreesToRadians } from "./rtweekend";
import { Image } from "./image";
import { Hittable, HitRecord } from "./hittable";
import { Interval } from "./interval";
import { Ray } from "./ray";
import { Color, Point3, Vec3, cross, randomInUnitDisk, unitVector } from "./vec3";

export class Camera {
  // Public config
  aspectRatio = 1.0;
  imageWidth = 100;
  samplesPerPixel = 10;
  maxDepth = 10;
  vfov = 90;
  lookFrom: Point3 = new Vec3([0, 0, 0]);
  lookAt: Point3 = new Vec3([0, 0, -1]);
  vup: Vec3 = new Vec3([0, 1, 0]);
  defocusAngle = 0;
  focusDist = 10;

  // Private
  private imageHeight = 0;
  private pixelSamplesScale = 1.0;
  private center: Point3 = this.lookFrom;
  private pixel00Loc: Point3 = this.lookFrom;
  private pixelDeltaU: Vec3 = new Vec3([0, 0, 0]);
  private pixelDeltaV: Vec3 = new Vec3([0, 0, 0]);
  private u: Vec3 = new Vec3([0, 0, 0]);
  private v: Vec3 = new Vec3([0, 0, 0]);
  private w: Vec3 = new Vec3([0, 0, 0]);
  private defocusDiskU: Vec3 = new Vec3([0, 0, 0]);
  private defocusDiskV: Vec3 = new Vec3([0, 0, 0]);

  initialize() {
    this.imageHeight = Math.floor(this.imageWidth / this.aspectRatio);
    if (this.imageHeight < 1) this.imageHeight = 1;
    this.pixelSamplesScale = 1.0 / this.samplesPerPixel;
    this.center = this.lookFrom;

    const theta = degreesToRadians(this.vfov);
    const h = Math.tan(theta / 2);
    const viewportHeight = 2 * h * this.focusDist;
    const viewportWidth = viewportHeight * (this.imageWidth / this.imageHeight);

    this.w = unitVector(this.lookFrom.sub(this.lookAt));
    this.u = unitVector(cross(this.vup, this.w));
    this.v = cross(this.w, this.u);

    const viewportU = this.u.mulScalar(viewportWidth);
    const viewportV = this.v.mulScalar(-viewportHeight);

    this.pixelDeltaU = viewportU.divScalar(this.imageWidth);
    this.pixelDeltaV = viewportV.divScalar(this.imageHeight);

    const viewportUpperLeft = this.center
      .sub(this.w.mulScalar(this.focusDist))
      .sub(viewportU.divScalar(2))
      .sub(viewportV.divScalar(2));
    this.pixel00Loc = viewportUpperLeft.add(this.pixelDeltaU.add(this.pixelDeltaV).mulScalar(0.5));

    const defocusRadius = this.focusDist * Math.tan(degreesToRadians(this.defocusAngle / 2));
    this.defocusDiskU = this.u.mulScalar(defocusRadius);
    this.defocusDiskV = this.v.mulScalar(defocusRadius);
  }

  private sampleSquare(): Vec3 {
    return new Vec3([Math.random() - 0.5, Math.random() - 0.5, 0]);
  }

  private defocusDiskSample(): Point3 {
    const p = randomInUnitDisk();
    return this.center
      .add(this.defocusDiskU.mulScalar(p.e[0]))
      .add(this.defocusDiskV.mulScalar(p.e[1]));
  }

  getRay(i: number, j: number): Ray {
    const offset = this.sampleSquare();
    const pixelSample = this.pixel00Loc
      .add(this.pixelDeltaU.mulScalar(i + offset.x()))
      .add(this.pixelDeltaV.mulScalar(j + offset.y()));
    let rayOrigin = this.center;
    if (this.defocusAngle > 0) {
      rayOrigin = this.defocusDiskSample();
    }
    const rayDirection = pixelSample.sub(rayOrigin);
    return new Ray(rayOrigin, rayDirection);
  }

  rayColor(r: Ray, depth: number, world: Hittable): Color {
    if (depth <= 0) return new Vec3([0, 0, 0]);
    const rec = new HitRecord();
    if (world.hit(r, new Interval(0.001, Number.POSITIVE_INFINITY), rec)) {
      const { didScatter, attenuation, scattered } = rec.mat.scatter(r, rec);
      if (didScatter) {
        return attenuation.mul(this.rayColor(scattered, depth - 1, world));
      }
      return new Vec3([0, 0, 0]);
    }
    const unitDirection = unitVector(r.direction());
    const a = 0.5 * (unitDirection.y() + 1.0);
    const white = new Vec3([1.0, 1.0, 1.0]);
    const blue = new Vec3([0.5, 0.7, 1.0]);
    return white.mulScalar(1.0 - a).add(blue.mulScalar(a));
  }

  renderSync(world: Hittable): string {
    this.initialize();
    const img = new Image(this.imageWidth, this.imageHeight);
    for (let j = 0; j < this.imageHeight; j++) {
      process.stdout.write(`\rScanlines remaining: ${this.imageHeight - j} `);
      for (let i = 0; i < this.imageWidth; i++) {
        let pixelColor = new Vec3([0, 0, 0]);
        for (let s = 0; s < this.samplesPerPixel; s++) {
          const r = this.getRay(i, j);
          pixelColor = pixelColor.add(this.rayColor(r, this.maxDepth, world));
        }
        img.setPixel(i, j, pixelColor.mulScalar(this.pixelSamplesScale));
      }
    }
    process.stdout.write(`\rScanlines remaining: 0 `);
    return img.toPPM();
  }
}


